Add write restriction middleware

talal opened the pull request:
Additionally, the sysmeta_domain_override middleware has been updated to remove legacy params/args.

talal committed:
Add gitignore

talal committed:
Add write restriction middleware

talal committed:
Improve README

talal committed:
Add write-restriction usage instructions

talal committed:
README: slight wording improvement

talal committed:
Update sysmeta-domain-override middleware

Remove legacy params/args that are no longer needed in newer Swift
versions.

talal committed:
Update class docstring

and add Limes reference.

majewsky commented on sapcc_swift_addons/write_restriction.py:
```diff
@@ -0,0 +1,102 @@
+
+        return self.app(env, start_response)
+
+    def write_restricted_container(self, req, account, container):
+        """
+        Check whether a container is write restricted, i.e. the
+        'X-Container-Meta-Write-Restricted' metadata header is set to 'true'.
+        """
+        info = get_info(self.app, req.environ, account, container, swift_source="RO")
+        write_restricted = info.get("meta", {}).get("write-restricted", "")
+        return config_true_value(write_restricted)
```
This should also match if the container is currently not write-restricted, but the current request would set the write restriction header to true. Non-privileged users should not be able to lock themselves out of their access.

reimannf replied:
Shouldn't we use an `X-Container-Sysmeta` prefix here. Than usual users are not allowed to set it anyhow.

reimannf replied:
Ignore that. We would need to call the container server directly

talal replied:
@majewsky I have updated the middleware such only allowed users can set/modify the `write-restricted` header regardless of whether the container is write restricted or not. This prevents user lock-out and also makes the `allowed_roles` option required instead of optional.

reimannf commented on sapcc_swift_addons/write_restriction.py:
```diff
@@ -0,0 +1,102 @@
+        Check whether a container is write restricted, i.e. the
+        'X-Container-Meta-Write-Restricted' metadata header is set to 'true'.
+        """
+        info = get_info(self.app, req.environ, account, container, swift_source="RO")
```
just a small thing: `swift_source` should be `WR` for write_restriction, not read only.

talal committed:
write_restriction: fix `swift_source` param

talal committed:
Make `allowed_roles` option required

so that users can not lock themselves out of their containers.

talal committed:
Fix user meta header check

talal committed:
Update error msg

Swift truncates the response message to first 60 characters therefore we
have to brief.

majewsky commented on sapcc_swift_addons/write_restriction.py:
```diff
@@ -0,0 +1,109 @@
+                return HTTPMethodNotAllowed(body=msg)(env, start_response)
+
+            if self.write_restricted_container(req, account, container):
+                msg = "Writes are disabled for this container."
```
This wording does not describe the situation accurately.

```suggestion
                msg = "Writes are restricted for this container."
```

majewsky reviewed (approved):
Thumbs up from my side, though I have one nitpick.

talal committed:
Improve error message
